### DISCUSSION TITLE
Meta-Tearing

### DISCUSSION THREAD

**Discussion by casella** [OpenModelica/OpenModelica#8480] [General]
I had an interesting discussion with a colleague about the solution of difficult steady-state modelling problems for thermal power generation systems, possibly including system design and component sizing.

What makes these problems really hard is the presence of recirculation loops. The trick is to tear the problem by cutting the loops open in appropriate places. This requires a domain modelling expert (hereinafter _the guru)_ who knows by experience and divine inspiration what are the best places to cut the loops, and which stream variables to use as tearing variables among the ones involved in the cut set  (pressure, mass flow rate, specific enthalpy, chemical composition) . 

That being done, the traditional way of solving the problem is to manually write down the solution procedure for each block of the system. This is done using some imperative language: Matlab, C++, Python, Java, C#, whatever. Every programmer has its favourite, but it doesn't really make any difference from a conceptual point of view. The important point is that, thanks to the _guru_'s tearing, these blocks can be executed sequentially. Then, some iterative strategy is used to eventualy converge to a system-wide solution.

The problem with the traditional approach is the manual writing of the blocks, and oftentimes also of the iterative strategy. First of all because it is tedious and error-prone. Most importantly, because the very same piece of equipment may require to write completely different sequential blocks times and again, depending on the specific choice of inputs and outputs. Last, but not least, the iterative strategy may be a bit primitive, not a state-of-the-art algorithm.

OpenModelica (and other Modelica tools) does use tearing. However, the torn variables are selected in a way that _they can all be solved sequentially by explicit assignments_. This is a very strong requirement, whereby a difficult problem, which may have thousands of unknowns, is reduced to a significantly smaller one, which however still has some hundreds of tearing variables. Initializing all of them to the right value to ensure convergence is doable (EDF does it all the time) but is really not practical. 

If you follow the _guru approach_, even the most difficult problems may have one or two dozens tearing variables in the worst case, for which the guru normally has some very definite idea about the initial guess values.

I think a smart equation-based approach could take the best of both worlds, with a very limited implementation effort. This is the idea.

The guru declares what are the tearing variable he knows are best to solve the problem in a robust way, e.g. by annotations. These annotations could be embedded in some ad-hoc components, so this can be done through a GUI. These variables are not taken as a suggestion by the tool, but rather as mandatory. 

The tool then determines the BLT of the system assuming these variables are known. In the best possible case, this BLT will be strictly lower triangular, which means standard tearing can be used. However, if you program the various blocks of the model following an a-causal approach, it is very likely that some algebraic loops will still be involved. The crucial point is that solving those loops should be straightforward from the point of view of convergence, given the wise choice of the top-level tearing variables by the guru. So, if the BLT of the torn system is not strictly lower triangular, the individual diagonal blocks corresponding to strong components can be subject to tearing again, in order to further reduce the complexity of the problem.

At the end of the day you get two nested layers of tearing. The top one has the tearing variables decided by the guru. The tool doesn't question them and just takes them as an input. The corresponding torn equations may still involve strong components, in which case the tool is free to use whatever strategy it deems fit to solve them, including using tearing again, using sparse solvers, basically everything we already have in place now when trying to solve a nonlinear initialization problem.

This strategy could be a game-changer in the field of energy systems, and could allow OpenModelica to challenge existing commercial tools such as Aspen, which have a big market. I see it as extremely promising, at least for applications in academia, where having more openness and flexibility is more important than having large libraries of models of existing stuff. BTW, we already have some very challenging test cases programmed in Modelica in this field, so testing this method in real-life conditions woud be fairly easy. I can take care of that myself.

The question is, how difficult would it be to prototype this approach in OMC?

@bernhardbachmann, @kabdelhak, @AnHeuermann, @phannebohm, can you comment on this?

**Comment by kabdelhak** [comment]
We have discussed this and i will start implementing the backend part. There are some structural changes that have to be done and i will assess how much work it is. If it is easy to do and viable, Andreas will take care of the runtime afterwards.

Could you provide a small example model on which we could test our implementation?

**Comment by kabdelhak** [comment]
I created a [branch](https://github.com/kabdelhak/OpenModelica/tree/7628_meta_tearing) regarding this issue where first stuff is done in the backend using as much information as possible. It fails during C compilation because of naming issues.

I used following naive simple model:
```
loadString("
  model M
    Real a(start = 10.0) annotation(tearingSelect=always),b,c(start = 12.0),d(start = 13.0);
  initial equation
    b = 20.0;
  equation
    a^2 + der(b) + c^2 = d;
    d = a^2 + der(b) + cos(time);
    c^3 * sin(der(b)) = a +d;
    cos(d) + cos(der(b)) + sin(a) = time;
  end M;
"); getErrorString();

simulate(M); getErrorString();
```

**Comment by phannebohm** [comment]
Can we please name it Guru-Tearing? I like how that sounds.

Also, @casella can you please clarify what exactly the runtime is supposed to do? My impression is that we would have big loops that (possibly) contain smaller loops. Then, for each iteration of one big loop, its small loops need to be solved first. Is that what you mean?

**Comment by kabdelhak** [comment]
> 
> 
> Can we please name it Guru-Tearing? I like how that sounds.
> 
> Also, @casella can you please clarify what exactly the runtime is supposed to do? My impression is that we would have big loops that (possibly) contain smaller loops. Then, for each iteration of one big loop, its small loops need to be solved first. Is that what you mean?

Currently it is called guruTearing in the Backend. :)

Also the nested loop idea is how i implemented it and generally i do not see any other option to enforce the users choice.

**Comment by casella** [comment]
> Could you provide a small example model on which we could test our implementation?

I can do it next week, if you please remind me.

**Comment by casella** [comment]
> Can we please name it Guru-Tearing? I like how that sounds.

That's fine for me 😄 
 
> Also, @casella can you please clarify what exactly the runtime is supposed to do? My impression is that we would have big loops that (possibly) contain smaller loops. Then, for each iteration of one big loop, its small loops need to be solved first. Is that what you mean?

Yes. The "loops" correspond to solving a nonlinear torn system by means of Newton's algorithm.

**Comment by casella** [comment]
> 
> 
> I created a [branch](https://github.com/kabdelhak/OpenModelica/tree/7628_meta_tearing) regarding this issue where first stuff is done in the backend using as much information as possible. It fails during C compilation because of naming issues.
> 
> I used following naive simple model:
> 
> ```
> loadString("
>   model M
>     Real a(start = 10.0) annotation(tearingSelect=always),b,c(start = 12.0),d(start = 13.0);
>   initial equation
>     b = 20.0;
>   equation
>     a^2 + der(b) + c^2 = d;
>     d = a^2 + der(b) + cos(time);
>     c^3 * sin(der(b)) = a +d;
>     cos(d) + cos(der(b)) + sin(a) = time;
>   end M;
> "); getErrorString();
> 
> simulate(M); getErrorString();
> ```

I'm not sure exactly how do you envision this model will be solved. What I understand is that if you use `a` as guru-tearing variable, you can first solve the second and fourth equations together for `der(b)` and `d` (possibly using tearing), then solve either the first or the third for `c`, and then use the other remaining equation for the residual of the outer tearing loop. Do I get it right?

Of course the big question here is how do you select which equation use for the residual, and which one(s) you keep for the possibly implicit torn part. This might be a big deal.

**Comment by kabdelhak** [comment]
> I'm not sure exactly how do you envision this model will be solved. What I understand is that if you use `a` as guru-tearing variable, you can first solve the second and fourth equations together for `der(b)` and `d` (possibly using tearing), then solve either the first or the third for `c`, and then use the other remaining equation for the residual of the outer tearing loop. Do I get it right?

To be honest i just fiddled around with it until it was solvable by the normal tearing but complicated enough to not get solved by other means.

> Of course the big question here is how do you select which equation use for the residual, and which one(s) you keep for the possibly implicit torn part. This might be a big deal.

That might be another question entirely. What i currently do is the following (taken from code comments):

```
  // ALGORITHM ABSTRACT
  //   1. get all equations and variables of the algebraic loop
  //   2. split variables in tearing and inner variables using TearingSelect.Always
  //   3. use existing matching information to extract residual equations
  //   4. manipulate existing matching information to represent the system without tearing variables and residual equations
  //   5. sort the manipulated system to get inner strong components
```

The idea is to just use the matching obtained earlier and use the equations that are matched to the selected tearing variables as residuals. After stripping tearing variables and residuals from the loop, the rest will just get sorted with tarjans algorithm again to obtain the ordering and strong components of inner equations (possibly containing other loops). Once this works we can think about enhancing this algorithm by making a smarter choice of residuals. Keep in mind that we usually do not choose the residuals, but instead we choose the inner equations such that they can be solved most effectively for inner variables. Everything leftover is considered residual.

**Comment by AnHeuermann** [comment]
@kabdelhak and myself are playing around with the current state of the guru tearing and encountered the following problem:

Let's say we have a model
```modelica
model guruTearing
  Real a(start=1) annotation(tearingSelect=always),b,c;
equation
  a^2 + b + c = sin(time);  // eq 1
  a + b + c^2 = tan(time); // eq 2
  a + b^2 + c = cos(time); // eq3
end guruTearing;
```

where we want to select `a`. So there are three options to choose the residual equation for the outer loop (eq1, eq2 and eq3).

So we set `a=1` and want to either solve
`System1`:
```
  a^2 + b + c = sin(time);  // eq 1
  a + b + c^2 = tan(time); // eq 2
```
or 
`System2`:
```
  a + b + c^2 = tan(time); // eq 2
  a + b^2 + c = cos(time); // eq3
```
or
`System3`:
```
  a^2 + b + c = sin(time);  // eq 1
  a + b^2 + c = cos(time); // eq3
```

There is no reason I can see why those systems should be solvable for some random `a`. When running WolframAlpha on those equation systems for `a=1` and `time=0` we have:
  - System 1: [Has a solution](https://www.wolframalpha.com/input/?i=%7Ba*a%2Bb%2Bc%3Dsin%280%29%2C+a%2Bb%2Bc%5E2%3Dcos%280%29%7D%2C+where+a+%3D1)
  - System 2: [No real solution](https://www.wolframalpha.com/input/?i=%7Ba%2Bb%2Bc%5E2%3Dsin%280%29%2C+a%2Bb%5E2%2Bc%3Dcos%280%29%7D%2C+where+a+%3D1)
  - System 3: [Has a solution](https://www.wolframalpha.com/input/?i=%7Ba*a%2Bb%2Bc%3Dsin%280%29%2C+a%2Bb%5E2%2Bc%3Dcos%280%29%7D%2C+where+a+%3D1)

And I don't see any reason why always at least one system of System1, System2 and System3 should be solvable at all for some initial value of `a`.

**Comment by phannebohm** [comment]
> And I don't see any reason why always at least one system of System1, System2 and System3 should be solvable at all for some initial value of `a`.

I would expect divine intervention here. There should be at least one system that's working, after all the guru chose `a` for a reason. But you have a point, maybe only a subset of possible systems actually works and just picking any residual is insufficient in general.

**Comment by casella** [comment]
> The idea is to just use the matching obtained earlier and use the equations that are matched to the selected tearing variables as residuals. After stripping tearing variables and residuals from the loop, the rest will just get sorted with tarjans algorithm again to obtain the ordering and strong components of inner equations (possibly containing other loops). Once this works we can think about enhancing this algorithm by making a smarter choice of residuals. 

OK. If I understand correctly, when dealing with strong components (which is what you are doing now), there is a certain amount of  arbitrariness in the matching, which implies that currently which variables are picked as residuals for the outer guru-loop is somewhat random. Fair enough, as a starting point.

> Keep in mind that we usually do not choose the residuals, but instead we choose the inner equations such that they can be solved most effectively for inner variables. Everything leftover is considered residual.

Sure, I get that. As I understand, the problem is that in this extended case, it is not too clear what "can be solved most effectively" actually means, since we may have strong components also in the outer (guru) loop inner equations, so it's not clear which ones we want to solve as inner variables and which ones to keep as residuals.

In fact, there would be one way to do this effectively: apply [Bachmann-Casella's Newton diagnostics](https://doi.org/10.1016/j.amc.2021.125991), using user's supplied start values 😃 

In this case, inner strong components characterized by low values of $alpha$, $Gamma$, and $sigma$ indicators of their iteration variables could be classified as "easily solved", because they are expected to converge without problems. We are talking about solving initialization problems here, not regular dynamic simulation problems, so guaranteeing convergence is a lot more important than efficiency, since the problem is only solved once.

The only problem I see with this is that it breaks the traditional boundaries between backend, codegen and runtime. You'd have to build partial solutions, generate the code for them, compute Jacobians and Hessians, and based on them obtain some indicators that can steer the symbolic manipulation in one way or another.

This may be an interesting problem for @JKRT, since I understand these boundaries are a lot more blurred in the Julia approach. Maybe we could have a brainstorming meeting with him when we have understood this thing a bit more?

**Comment by casella** [comment]
> > And I don't see any reason why always at least one system of System1, System2 and System3 should be solvable at all for some initial value of `a`.
> 
> I would expect divine intervention here. There should be at least one system that's working, after all the guru chose `a` for a reason. But you have a point, maybe only a subset of possible systems actually works and just picking any residual is insufficient in general.

Never underestimate the guru's karma 😄 

Seriously, what I expect to happen in real problems is that the guru's choice will end up leading to a lot of explicitly solvable inner equations, and a comparatively low number or remaining strong components.  But I understand you got to the same point I got in the previous comment: it is still not clear what criterion should we use to decide which implicit equations to keep for the outer loop, and which ones to solve as inner equations.

I understand that currently you have some criteria, see the (in)famous `--tearingStricness` flag, to do that. We need to extend them adding criteria for strong components. As I wrote in the previous comment, what matters is the likelyhood of convergence. The problem is that this convergence may depend on numerical values of some variables or parameters that were previusly computed in the global BLT, and with the current OMC architecture, there's no way you can get those values when you would need them.

Maybe we could start with some simpler heuristic criterion to judge which nonlinear systems are "more nonlinear" than others, and thus potentially more prone to convergence failure if not properly initialized. If we have no idea, we can always use `casualTearing`, we already have this feature in regular tearing 😄 

Remember that the overarching goal of this thing is that the guru should bother about picking a few strategic start values, obtained by his inner third eye insight, but then everything else should more or less work out of the box with default provided start values. At least, this is the scenario we have in mind.

**Comment by casella** [comment]
I guess a more realistic example case is due. I can try to provide it next week, but I need to get through some exam papers and a master's thesis review first.

In the meantime, you may want to have a second look at [#6196](https://trac.openmodelica.org/OpenModelica/ticket/6196#comment:15), which is somewhat related to this issue and is a serious nuisance for AC power systems.
